colcon_sanitizer_reports/sanitizer_log_parser.py

```diff
@@ -0,0 +1,195 @@
+from collections import defaultdict
+import csv
+from io import StringIO
+import re
+from typing import Dict, List, Optional, Tuple
+
+
+class _KeysFactory:
+    """Finds key stack trace lines in sanitizer error subsections."""
+
+    @staticmethod
+    def _make_masked_line(line: str) -> str:
+        """Masks numbers in a sanitizer error line so that keys from similar errors will match."""
+        masked_line = line
+        masked_line = re.sub(r'0x[\dabcdef]+', '0xX', masked_line)
+        masked_line = re.sub(r' \d+ ', ' X ', masked_line)
+        masked_line = re.sub(r'#\d+', '#X', masked_line)
+
+        return masked_line
+
+    @staticmethod
+    def _make_keys_base(*, lines: Tuple[str], stack_begin_regexes: Tuple[str, ...]) -> Tuple[str]:
```

Comment:
This function is a bit complex, could this be split into smaller ones focused on "one thing" (going through the logs, applying a regexp search, etc.)